tests/unit/s2n_self_talk_ktls_test.c

```diff
@@ -0,0 +1,250 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+#include <fcntl.h>
+#include <netinet/in.h>
+#include <string.h>
+#include <sys/socket.h>
+#include <sys/types.h>
+#include <unistd.h>
+
+#include "s2n_test.h"
+#include "testlib/s2n_testlib.h"
+#include "tls/s2n_ktls.h"
+#include "utils/s2n_random.h"
+
+/* There are issues with MacOS and FreeBSD so we define the constant ourselves.
+ * https://stackoverflow.com/a/34042435 */
+#define S2N_TEST_INADDR_LOOPBACK 0x7f000001 /* 127.0.0.1 */
+
+/* Unlike our other self-talk tests, this test cannot use AF_UNIX / AF_LOCAL.
+ * For a real self-talk test we need real kernel support for kTLS, and only
+ * AF_INET sockets support kTLS.
+ */
+static S2N_RESULT s2n_new_inet_socket_pair(struct s2n_test_io_pair *io_pair)
+{
+    RESULT_ENSURE_REF(io_pair);
+
+    int listener = socket(AF_INET, SOCK_STREAM, 0);
+    RESULT_ENSURE_GT(listener, 0);
+
+    struct sockaddr_in saddr = { 0 };
+    saddr.sin_family = AF_INET;
+    saddr.sin_addr.s_addr = htonl(S2N_TEST_INADDR_LOOPBACK);
+    saddr.sin_port = 0;
+
+    socklen_t addrlen = sizeof(saddr);
+    RESULT_ENSURE_EQ(bind(listener, (struct sockaddr *) &saddr, addrlen), 0);
+    RESULT_ENSURE_EQ(getsockname(listener, (struct sockaddr *) &saddr, &addrlen), 0);
```

Comment:
Shouldn't you do something with the output of this call? Otherwise why call the getter at all?

Reply:
The output is saddr, and we use it on line 60. Because we bind the server to port 0, the os assigns the server a random unused port, and we need getsockname to tell us what port the server was assigned so that the client can connect to it.